File: src/sliding-windows/sliding_windows_test.go
package slidingwindows_test

import (
	slidingwindows "algorithm-puzzle-exercise/src/sliding-windows"
	"reflect"
	"testing"
)

func TestFindAnagrams(t *testing.T) {
	testCases := []struct {
		s        string
		target   string
		expected []int
	}{
		{"cbaebabacd", "abc", []int{0, 6}},
		{"abab", "xyz", []int{}},
		{"a", "ab", []int{}},
		{"abacbabc", "abc", []int{1, 2, 3, 5}},
		{"aaaaa", "aa", []int{0, 1, 2, 3}},
	}

	for _, tc := range testCases {
		result := slidingwindows.FindAnagram(tc.s, tc.target)
		if !reflect.DeepEqual(result, tc.expected) {
			t.Errorf("FindAnagrams(%q, %q) = %v; want %v", tc.s, tc.target, result, tc.expected)
		}
	}
}

func TestLongestSubstringWithUniqueChar(t *testing.T) {
	cases := []struct {
		s    string
		want int
	}{
		{"", 0},
		{"abcabcbb", 3},
		{"bbbbb", 1},
		{"pwwkew", 3},
		{"au", 2},
	}
	for _, c := range cases {
		if got := slidingwindows.LongestSubstringWithUniqueChar(c.s); got != c.want {
			t.Errorf("got %d for %q, want %d", got, c.s, c.want)
		}
	}
}

File: src/sliding-windows/sliding-windows.test.ts
import { describe, it, expect } from 'vitest'
import { longestSubstringWithUniqueChar } from './longest-substring-with-unique-char'
import { findAnagrams } from './substring-anagram'

describe('findAnagrams', () => {
    it('should find anagrams in a standard case', () => {
        expect(findAnagrams('cbaebabacd', 'abc')).toEqual([0, 6])
    })

    it('should return empty array if no anagrams are found', () => {
        expect(findAnagrams('abab', 'xyz')).toEqual([])
    })

    it('should return empty array if s is shorter than target', () => {
        expect(findAnagrams('a', 'ab')).toEqual([])
    })

    it('should find multiple overlapping anagrams', () => {
        expect(findAnagrams('abacbabc', 'abc')).toEqual([1, 2, 3, 5])
    })

    it('should handle strings with all same characters', () => {
        expect(findAnagrams('aaaaa', 'aa')).toEqual([0, 1, 2, 3])
    })
})

describe('longestSubstringWithUniqueChar', () => {
    const cases: [string, number][] = [
        ['', 0],
        ['abcabcbb', 3],
        ['bbbbb', 1],
        ['pwwkew', 3],
        ['au', 2],
    ]
    cases.forEach(([s, want]) => {
        it(`${s}->${want}`, () => {
            expect(longestSubstringWithUniqueChar(s)).toBe(want)
        })
    })
})
